Linking error for htslib 1.17+

We originally built this special Windows build of htslib with version 1.15.1. Recently I was able to build htslib 1.16 with minimal effort (https://github.com/TileDB-Inc/m2w64-htslib-build/pull/3). I only had to delete an outdated patch (ie we had backported a future change to 1.15.1).

However, I get the same build error when I try to build 1.17, 1.18, or 1.19. Thus we won't be able to build tiledb-vcf-feedstock against a more recent version of htslib on Windows until this is resolved.

## The error

I think the error has to do with linking zlib:

```
gcc -fvisibility=hidden  -o test/test_faidx test/test_faidx.o libhts.a -lz /mingw64/lib/libregex.a /mingw64/lib/libsystre.a /mingw64/lib/libtre.a /mingw64/lib/libintl.a /mingw64/lib/libiconv.a /mingw64/lib/libdeflate.a /mingw64/lib/liblzma.a /mingw64/lib/libbz2.a /mingw64/lib/libcrypt32.a /mingw64/lib/libwinpthread.a /mingw64/lib/libcurl.a /mingw64/lib/libcrypto.a /mingw64/lib/libssl.a /mingw64/lib/libwldap32.a /mingw64/lib/libcrypto.a /mingw64/lib/libcrypt32.a /mingw64/lib/libcertcli.a /mingw64/lib/libzstd.a /mingw64/lib/libbrotlidec.a /mingw64/lib/libbrotlienc.a /mingw64/lib/libbrotlicommon.a /mingw64/lib/libssh2.a /mingw64/lib/libidn2.a /mingw64/lib/libpsl.a -lws2_32 /mingw64/lib/libunistring.a /mingw64/lib/libnghttp2.a /mingw64/lib/libz.a /mingw64/lib/libbcrypt.a  -lpthread
D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/a/_temp/msys64/mingw64/lib/libz.a(deflate.o):(.text+0x1f20): multiple definition of `deflate'; D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../lib/libz.dll.a(zlib1_dll_d000023.o):(.text+0x0): first defined here
D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/a/_temp/msys64/mingw64/lib/libz.a(deflate.o):(.text+0x33c0): multiple definition of `deflateEnd'; D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../lib/libz.dll.a(zlib1_dll_d000026.o):(.text+0x0): first defined here
D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/a/_temp/msys64/mingw64/lib/libz.a(deflate.o):(.text+0x3470): multiple definition of `deflateInit2_'; D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../lib/libz.dll.a(zlib1_dll_d000028.o):(.text+0x0): first defined here
D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/a/_temp/msys64/mingw64/lib/libz.a(inflate.o):(.text+0x230): multiple definition of `inflateReset'; D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../lib/libz.dll.a(zlib1_dll_d000087.o):(.text+0x0): first defined here
D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/a/_temp/msys64/mingw64/lib/libz.a(inflate.o):(.text+0x340): multiple definition of `inflateInit2_'; D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../lib/libz.dll.a(zlib1_dll_d000083.o):(.text+0x0): first defined here
D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/a/_temp/msys64/mingw64/lib/libz.a(inflate.o):(.text+0x4e0): multiple definition of `inflate'; D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../lib/libz.dll.a(zlib1_dll_d000074.o):(.text+0x0): first defined here
D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/a/_temp/msys64/mingw64/lib/libz.a(inflate.o):(.text+0x22a0): multiple definition of `inflateEnd'; D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../lib/libz.dll.a(zlib1_dll_d000080.o):(.text+0x0): first defined here
collect2.exe: error: ld returned 1 exit status
make: *** [Makefile:638: test/test_faidx] Error 1
```

## Troubleshooting

As expected, the line that produces the error was added to the `Makefile` between 1.16 and 1.17 ([diff](https://github.com/samtools/htslib/compare/1.16...1.17)) in PR https://github.com/samtools/htslib/pull/1519

Frustratingly, the Windows MSYS2/MINGW64 build in `.appveyor.yml` hasn't been touched since 2022-02-04 ([commit](https://github.com/samtools/htslib/commit/3f7e13e0688e50fca61e6f043e6655a165d596a2)), ie it hasn't changed since htslib 1.15.

I also looked at the build recipe for [mingw-w64-htslib](https://github.com/msys2/MINGW-packages/commits/master/mingw-w64-htslib/PKGBUILD). They've made various changes between 1.15.1 and 1.17 ([diff](https://github.com/msys2/MINGW-packages/compare/bef8381316ceb5154c12c7f09b574e61027c8c59...72399b72963228ab953db2ae7a052dee635abcbb#diff-7e2a950b8b656c4fed70d6054736843526bfedf2cabd8d22b5ca42fdca7c4d56)), but I don't see anything that would directly address this zlib linking error. I tried explicitly installing `mingw-w64-x86_64-libwinpthread`
like they did, but that made no difference.

```
gcc … -o test/test_faidx test/test_faidx.o libhts.a -lz … /mingw64/lib/libz.a …
ld.exe: D:/a/_temp/msys64/mingw64/lib/libz.a(deflate.o):(.text+0x1f20): multiple definition of `deflate';
D:/a/_temp/msys64/mingw64/lib/libz.dll.a(zlib1_dll_d000023.o):(.text+0x0): first defined here
```

The problem is that these test executables are being linked to libz both dynamically and statically. Presumably all that `/mingw64/lib/libz.a` etc is something you have intentionally configured into `$(LIBS)` so that the resulting Windows libraries will be statically linked.

So you probably want to remove the `-lz` from the _test/test_foo_ recipes that contain it explicitly. To be honest, they probably shouldn't be there upstream either, as an appropriate form of `-lz` will always be provided by `$(LIBS)`.

--- a/.github/workflows/msys2-htslib-build.yml
+++ b/.github/workflows/msys2-htslib-build.yml
@@ -12,7 +12,7 @@ on:
       - 'scripts/**'
   workflow_dispatch:
 env:
-  PACKAGE_VERSION: 1.16
+  PACKAGE_VERSION: 1.17
   LIBHTS_SOVERSION: 3
   RELEASE_VERSION: 0 # equivalent to conda build number
 jobs:

--- a/patches/makefile.staticlink.patch
+++ b/patches/makefile.staticlink.patch
@@ -1,52 +1,56 @@
-574c574
+632c632
 < 	$(CC) $(LDFLAGS) -o $@ test/test_bgzf.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/test_bgzf.o libhts.a  $(LIBS) -lpthread
-577c577
+635c635
 < 	$(CC) $(LDFLAGS) -o $@ test/test_expr.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/test_expr.o libhts.a  $(LIBS) -lpthread
-580c580
+638c638
+< 	$(CC) $(LDFLAGS) -o $@ test/test_faidx.o libhts.a -lz $(LIBS) -lpthread
+---
+> 	$(CC) $(LDFLAGS) -o $@ test/test_faidx.o libhts.a $(LIBS) -lpthread
+641c641
 < 	$(CC) $(LDFLAGS) -o $@ test/test_kfunc.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/test_kfunc.o libhts.a  $(LIBS) -lpthread
-583c583
+644c644
 < 	$(CC) $(LDFLAGS) -o $@ test/test_kstring.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/test_kstring.o libhts.a  $(LIBS) -lpthread
-613c613
+677c677
 < 	$(CC) $(LDFLAGS) -o $@ test/test-bcf-sr.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/test-bcf-sr.o libhts.a  $(LIBS) -lpthread
-616c616
+680c680
 < 	$(CC) $(LDFLAGS) -o $@ test/test-bcf-translate.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/test-bcf-translate.o libhts.a  $(LIBS) -lpthread
-698c698
+768c768
 < 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads1.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads1.o libhts.a  $(LIBS) -lpthread
-701c701
+771c771
 < 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads2.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads2.o libhts.a  $(LIBS) -lpthread
-704c704
+774c774
 < 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads3.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads3.o libhts.a  $(LIBS) -lpthread
-707c707
+777c777
 < 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads4.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads4.o libhts.a  $(LIBS) -lpthread
-710c710
+780c780
 < 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads5.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads5.o libhts.a  $(LIBS) -lpthread
-713c713
+783c783
 < 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads6.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads6.o libhts.a  $(LIBS) -lpthread
-716c716
+786c786
 < 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads7.o libhts.a -lz $(LIBS) -lpthread
 ---
 > 	$(CC) $(LDFLAGS) -o $@ test/thrash_threads7.o libhts.a  $(LIBS) -lpthread

